soc: mps2_an385: Added support for MPU

AdithyaBaglody opened the pull request:
Added MPU support for MPS2_AN385

AdithyaBaglody committed:
soc: mps2_an385: Added support for MPU

Added required files/support for the MPU in MPS2-AN358 (Cortex-M3).

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

AdithyaBaglody committed:
soc: mps2_an385: Enable Memory Protection unit.

This patch will enable the MPU in MPS2_AN385 soc.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

AdithyaBaglody committed:
tests: userspace: Update the required stack size for mps2_an385

Increasing the stack size to 1024.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

galak commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -43,6 +43,13 @@ static inline u32_t _size_to_mpu_rasr_size(u32_t size)
 		return REGION_32B;
 	}
 
+#ifdef CONFIG_SOC_SERIES_MPS2
```
This is because of QEMU, not the actual hardware for MPS2_AN385

andrewboie replied:
Could we perhaps make this a Kconfig instead? And set to 1024 in the qemu_cortex_m3 defconfig

AdithyaBaglody replied:
How can I inform the build system that I am using qemu rather than a real hardware.
Is that I need to create a new board for this, say qemu_mps2_an385?

andrewboie replied:
no, just change qemu_cortex_m3 to use this SOC definition instead of the other one.
And then ifdef this based on BOARD=qemu_cortex_m3

AdithyaBaglody replied:
This causes build failures on all Bluetooth related tests/samples. This change would be backward incompatible.
So another solution may be something like BOARD=qemu_cortex_m3_mpu

galak reviewed (changes requested):
The various places that test against SOC_SERIES_MPS2 aren't correct, the actual hw is fine, its QEMU that has the 1k limitation.

andrewboie commented on boards/arm/mps2_an385/mps2_an385_defconfig:
```diff
@@ -28,3 +28,10 @@ CONFIG_UART_CMSDK_APB=y
+#userspace
+CONFIG_ARCH_HAS_USERSPACE=y
+CONFIG_ARM_MPU=y
+CONFIG_MAIN_STACK_SIZE=2048
```
Why do we need to override the main/idle/privileged stack size from defaults?

AdithyaBaglody replied:
Qemu cant handle stacks lesser than 1024.  This will be a region i need to congfigure in the MPU and the qemu has a 1k limitation.

andrewboie replied:
then why aren't you changing this in qemu_cortex_m3 board definition?

andrewboie replied:
These stack size changes need to be moved out of the mps2_an385 configuration, I can't merge this since this is really something that affects QEMU and not the actual mps2_an385 board.

nashif replied:
@AdithyaBaglody you never addressed the comment above.

AdithyaBaglody replied:
oh I had actually addressed it before. Now after enabling the userspace(in the last 2days) this came back in.
will fix it. Thanks for catching this.

andrewboie commented on scripts/gen_alignment_script.py:
```diff
@@ -63,10 +63,17 @@ def main():
+
     if bit_len:
-        align_size = 1 << bit_len
+        if "CONFIG_SOC_SERIES_MPS2" in syms and bit_len < 10:
```
yeah another hack here -- let's just pull this out of kconfig

andrewboie commented on tests/kernel/mem_protect/userspace/src/main.c:
```diff
@@ -324,7 +324,11 @@ static void pass_noperms_object(void)
 __kernel struct k_thread kthread_thread;
 
+#ifdef CONFIG_SOC_SERIES_MPS2
+#define STACKSIZE 1024
```
just make this 1024 for everyone IMO, 512 was selected at random IIRC and it's not like MPU-enabled devices have really tiny amounts of RAM

agross-linaro replied:
i agree.  i always wondered about that.  If people want to minimize that stuff they can add pare it back down sub <1k.

AdithyaBaglody committed:
drivers: timers: systick: Enforce a barrier in _timer_cycle_get_32.

The code assumes that when the systick counter hits zero,
the timer interrupt will be taken before the loop can
read the LOAD/VAL registers, but this is not architecturally
guaranteed, and so the code can see a post-reload SysTick->VAL
and a pre-reload clock_accumulated_count, which causes it to
return an incorrectly small cycle count. By adding a ISB we
overcome this issue.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

andrewboie commented on scripts/gen_alignment_script.py:
```diff
@@ -62,11 +62,15 @@ def main():
+            align_size = 1 << bit_len
     else:
-        align_size = 32
+            align_size = min_region_size
```
indentation is incorrect here.

tyrantlizard77 reviewed (changes requested):
Getting close, but a few changes still needed.

Any details on the bluetooth issues you are seeing?

galak commented on arch/arm/soc/arm/mps2/arm_mpu_regions.c:
```diff
@@ -0,0 +1,24 @@
+#include <soc.h>
+#include <arch/arm/cortex_m/mpu/arm_mpu.h>
+
+#define FLASH_BASE_ADDRESS (0x00000000)
```
Why aren't these coming from DT?

AdithyaBaglody replied:
we do not have a DTS file for mps2_an385. So currently this will be kept here until we have one.

nashif replied:
> we do not have a DTS file for mps2_an385. So currently this will be kept here until we have one.

boards/arm/mps2_an385/mps2_an385.dts ???

AdithyaBaglody committed:
tests: kernel: pipe : Update the stack size.

Now the stack size is a function of CONFIG_TEST_EXTRA_STACKSIZE.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: mem_protect: mem_protect: Update the stack size.

This patch updates the alignment for the memory domain partitions.
Also update the stack size for qemu_cortex_m3.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

tyrantlizard77 reviewed (changes requested):
> Dependencies: QEMU version 2.12

This should NOT be dependent on QEMU 2.12.
Separate out any work in QEMU 2.12 into another PR.
The mps3_an385 is a real piece of hardware with a working MPU.
Issues with QEMU are separate and need to be dealt with separately.

AdithyaBaglody commented on boards/arm/mps2_an385/mps2_an385_defconfig:
```diff
@@ -28,3 +28,8 @@ CONFIG_UART_CMSDK_APB=y
+#userspace
+CONFIG_MAIN_STACK_SIZE=2048
+CONFIG_IDLE_STACK_SIZE=1024
+CONFIG_PRIVILEGED_STACK_SIZE=1024
```
we have quite a lot of memory on the actual board. So if someone wants to run it on qemu he shouldn't face the issue i did when i started out.

nashif replied:
ditto

AdithyaBaglody commented on samples/mpu/mem_domain_apis_test/src/main.c:
```diff
@@ -10,7 +10,7 @@
 
 /* size of stack area used by each thread */
-#define STACKSIZE 1024
+#define STACKSIZE CONFIG_MAIN_STACK_SIZE
```
This is one of the smiplest ways to get it to work on qemu. this sample needs 2k of stack and i didn't want to do a #if def CONFIG_BOARD_QEMU_m3. Maybe i will end up doing just that.

AdithyaBaglody commented on tests/benchmarks/sys_kernel/testcase.yaml:
```diff
@@ -1,5 +1,6 @@
 tests:
   benchmark.kernel:
     arch_exclude: nios2 riscv32 xtensa
+    platform_exclude: qemu_cortex_m3
```
the prj.conf streches the time between timer interrupts. This on qemu will overflow the counter.
And what is the point in running a benchmark on an emulation platform?

AdithyaBaglody replied:
>Why is this not a problem on any of our other emulated platforms?

In arm we have only 24 bits to represent the count value. This maybe one reason. The other the clock speed on other platforms might be lesser. So it ends up streaching the timer Interrupts according to the prj.conf.

>Because the test still reports success or failure, and sanitycheck only runs on emulated platforms. If there is some functional issue that causes this test to crash we need to know about it.

As mentioned above, this is to do with the timer interrupts. One workaround would be to create a new prj.conf for qemu_cortex_m3.

andrewboie commented on samples/philosophers/src/main.c:
```diff
@@ -83,7 +83,7 @@
 /***************************************/
 
-#define STACK_SIZE 768
+#define STACK_SIZE 1024
```
I'm trying to understand something here.
IIRC, On ARM, the guard region for stacks is taken from the existing stack size. It is not added like on x86.
Wouldn't a stack size of 1024 be entirely consumed by the guard area, leaving no room for the actual stack buffer?

AdithyaBaglody replied:
yes. if the MPU guard is enabled it will consume 1K out of the stack size.
but I want to create a warning that will inform the users of this limitation when we refactor qemu_cortex_m3.

AdithyaBaglody replied:
MPU Stack guard is not enabled by default. Hence this works.

andrewboie replied:
MPU stack guard is now enabled by default for tests.
We don't have the 1K limitation any more, correct?

nashif commented on tests/benchmarks/sys_kernel/testcase.yaml:
```diff
@@ -3,3 +3,7 @@ tests:
     arch_exclude: nios2 riscv32 xtensa
     min_ram: 32
     tags: benchmark
+  benchmark.kernel.qemu_cortex_m3:
```
not sure how this is related to this PR.

AdithyaBaglody replied:
Some  of these patches are need to keep the sanity when we move to the new QEMU. 
So such patches need to be ready before we update the SDK with the new QEMU.

nashif commented on tests/benchmarks/sys_kernel/prj_1.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_TEST=y
```
do not use number in file names, call it with something meaningful

andrewboie commented on samples/mpu/mem_domain_apis_test/src/main.c:
```diff
@@ -10,7 +10,11 @@
 #include <misc/printk.h>
 
 /* size of stack area used by each thread */
+#if (CONFIG_MAIN_STACK_SIZE > 1024)
```
Do we still need this?

andrewboie commented on samples/mpu/mem_domain_apis_test/src/main.c:
```diff
@@ -103,21 +107,21 @@ void main(void)
 	/* create application threads */
 	k_thread_create(&app_thread_id[0], app_stack[0],
-			K_THREAD_STACK_SIZEOF(app_stack[0]), app_thread,
+			STACKSIZE, app_thread,
```
why do this here? K_THREAD_STACK_SIZEOF() of the stack object in use should be sufficient.

AdithyaBaglody replied:
This was causing some issue on the MPU 1k alignment. This was leading to a stacking error.
So this stack sizeof returns a size - mpu guard size. Then inside the _new _thread API it will again subtract the size with the guard size. This was effectively making the stack size to become 0 in this case.

galak replied:
Have we tried newer qemu, I think support for smaller than 1k size might be supported now.

galak replied:
> Have we tried newer qemu, I think support for smaller than 1k size might be supported now.

Possibly in a 3.0.0-rc release.

andrewboie commented on samples/philosophers/sample.yaml:
```diff
@@ -4,6 +4,7 @@ common:
   extra_args: "-DDEBUG_PRINTF=1"
   tags: samples
   harness: console
+  min_ram: 32
```
Why do we need this now?
This used to work on even pretty small devices.

andrewboie commented on arch/arc/core/mpu/Kconfig:
```diff
@@ -38,3 +38,10 @@ config ARC_MPU
 	help
 	  Target has ARC MPU (currently only works for EMSK 2.2/2.3 ARCEM7D)
+
+config MPU_MIN_REGION_SIZE
```
I don't think we need this stuff any more

andrewboie commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -52,6 +47,12 @@ static inline u32_t _size_to_mpu_rasr_size(u32_t size)
 	}
 
+	/* The minimal supported region size as defined by the kconfig */
+	if (size <= CONFIG_MPU_MIN_REGION_SIZE) {
```
Think we don't need this change any more

tyrantlizard77 reviewed (dismissed):
Lot of leftover stuff from when QEMU only supported 1K+ regions, let's get rid of it

AdithyaBaglody committed:
tests: userspace: Incorrect location to the privileged stack.

The test case was supposed to access the privileged stack area
but instead it was accessing the stack guard region.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: benchmarks: sys_kernel: Reduce the number of loops.

The total number of loops being executed were taking far too long
to finish executing the test case. This was causing the timer handler
to execute in the middle of the test case. Thereby causing a test case
failure.
Now the number of loops have been reduced and this will make sure the
test case completes.

This change has no effect on the time being calculated by the
benchmark.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
include: arm: CMSIS header redefining MPU_PRESENT.

The header should check if the macro _MPU_PRESENT is defined
and create it only if not defined.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody commented on tests/kernel/mem_protect/mem_protect/src/mem_protect.h:
```diff
@@ -25,25 +25,17 @@ extern bool valid_fault;
 
 /* for kobject.c */
-#ifdef CONFIG_BOARD_MPS2_AN385
+#ifdef CONFIG_BOARD_QEMU_CORTEX_M3
```
No we dont. Removing this.

AdithyaBaglody committed:
boards: mps2_an385: Enable userspace for MPS2_AN385.

This patch will enable the userspace.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody commented:
@nashif please review.

nashif commented on boards/arm/mps2_an385/mps2_an385_defconfig:
```diff
@@ -28,3 +28,11 @@ CONFIG_UART_CMSDK_APB=y
 CONFIG_I2C=y
+
+#Enable MPU
+CONFIG_ARCH_HAS_USERSPACE=y
```
that is not needed and is automatically selected.

AdithyaBaglody replied:
This was left over code that came from the qemu enabling patches. Removed it.

nashif commented:
@AdithyaBaglody 2 tests are failing for me:

tests/kernel/mem_pool/mem_pool_threadsafe/kernel.memory_pool
and
tests/kernel/mem_protect/syscalls/kernel.memory_protection.syscalls

AdithyaBaglody commented:
>@AdithyaBaglody 2 tests are failing for me:
tests/kernel/mem_pool/mem_pool_threadsafe/kernel.memory_pool
and
tests/kernel/mem_protect/syscalls/kernel.memory_protection.syscalls

Anas can you please check with this rebase. I am not seeing this issue on my end.

nashif commented:
tests/kernel/mem_pool/mem_pool_threadsafe still fails, the syscalls one is probably expected to fault, we see this in other places. @dcpleung

dcpleung commented:
I ran sanitycheck with QEMU 3.0.0 couple times in a row. Every time it fails of different set of tests. Maybe it's a QEMU issue?

nashif commented:
> I ran sanitycheck with QEMU 3.0.0 couple times in a row. Every time it fails of different set of tests. Maybe it's a QEMU issue?

i am running qemu from head, so post 3.0.0

AdithyaBaglody commented:
@dcpleung You need to take the latest code i.e the repo head. Needed for MPU to work correctly. 
@nashif I am using a qemu that is 1/2 week old. There it seems to work just fine. Let me check with the latest repo head.

dcpleung commented:
I grabbed QEMU from git this morning (v3.0.0-614-g19b599f766), and it still gave me different set of failures between sanitycheck runs. Some of them are related to timers (e.g. too many ticks elapsed waiting for a task). However, I got a few USAGE FAULTs complaining about illegal use of EPSR and a few BUS FAULTS also.

AdithyaBaglody commented:
So if you want to run it on qemu. you need to add the following to the defconfig
CONFIG_MAIN_STACK_SIZE=1024
CONFIG_IDLE_STACK_SIZE=1024
CONFIG_PRIVILEGED_STACK_SIZE=1024
CONFIG_TEST_EXTRA_STACKSIZE=1024

nashif commented:
@AdithyaBaglody this now only failed on one test:


samples/basic/userspace/shared_mem

```
QEMU: ***** Booting Zephyr OS zephyr-v1.13.0-146-gaa839c2245 *****
QEMU: init partitions complete
QEMU: init app memory complete
QEMU: ENC Thread Created 200008b0
QEMU: init domain complete
QEMU: Partitions added to dom1
QEMU: dom1 Created
QEMU: PT Thread Created 20000a00
QEMU: dom0 Created
QEMU: CT Thread Created 20000974
QEMU: dom2 Created
QEMU: ***** MPU FAULT *****
QEMU: Data Access Violation
QEMU: MMFAR Address: 0x200000a0
QEMU: ***** Hardware exception *****
QEMU: Current thread ID = 0x200008b0
QEMU: Faulting instruction address = 0x4f6
QEMU: Fatal fault in thread 0x200008b0! Aborting.
QEMU: ENC thread started
QEMU: PT thread started
QEMU complete (timeout) after 60.059631 seconds
MAKE: 'sanity_test_finished mps2_an385/samples/basic/userspace/shared_mem/kernel.memory_protection.userspace'

```

AdithyaBaglody commented:
@nashif this was not a problem on qemu. It seems like it caught an issue which was unnoticed till now.
Created a PR for the solution #10087

nashif commented on tests/benchmarks/sys_kernel/prj_qemu_cortex_m3.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_TEST=y
```
Please remove this patch from this PR, not related to mps2_an385

nashif commented on tests/benchmarks/sys_kernel/testcase.yaml:
```diff
@@ -6,4 +6,4 @@ tests:
     extra_args: CONF_FILE=prj_qemu_cortex_m3.conf
-    tags: benchmark qemu_cortex_m3
\ No newline at end of file
+    tags: benchmark qemu_cortex_m3
```
remove this change to this file, not sure why we have qemu_cortex_m3 in the tags
